File: src/types/videoGenerate.ts
import { z } from "zod";

// ===== Slide Content Types =====

export const SlideContentSchema = z.object({
  title: z.string(),
  body: z.string().default(""),
  bulletPoints: z.array(z.string()).default([]),
  durationHint: z.number().nullable().optional(),
});
export type SlideContent = z.infer<typeof SlideContentSchema>;

// ===== Configuration Types =====

export const VideoThemeSchema = z.enum(["light", "dark"]);
export type VideoTheme = z.infer<typeof VideoThemeSchema>;

export const VideoTransitionSchema = z.enum(["cut", "fade"]);
export type VideoTransition = z.infer<typeof VideoTransitionSchema>;

export const VideoConfigSchema = z.object({
  width: z.number().default(1920),
  height: z.number().default(1080),
  theme: VideoThemeSchema.default("light"),
  transition: VideoTransitionSchema.default("cut"),
  title: z.string().nullable().optional(),
});
export type VideoConfig = z.infer<typeof VideoConfigSchema>;

export const VideoTTSConfigSchema = z.object({
  provider: z.string(),
  voice: z.string(),
  apiKey: z.string().nullable().optional(),
  baseUrl: z.string().nullable().optional(),
  model: z.string().nullable().optional(),
  speed: z.number().nullable().optional(),
});
export type VideoTTSConfig = z.infer<typeof VideoTTSConfigSchema>;

// ===== Result Types =====

export const VideoGenerationResultSchema = z.object({
  videoPath: z.string(),
  durationSeconds: z.number(),
  slideCount: z.number(),
  generationTimeSeconds: z.number(),
});
export type VideoGenerationResult = z.infer<typeof VideoGenerationResultSchema>;

// ===== Availability Types =====

export const VideoAvailabilitySchema = z.object({
  pillow: z.boolean(),
  ffmpeg: z.boolean(),
  pydub: z.boolean(),
  fullyAvailable: z.boolean(),
});
export type VideoAvailability = z.infer<typeof VideoAvailabilitySchema>;

// ===== Aspect Ratio Presets =====

export interface AspectRatioPreset {
  id: string;
  name: string;
  width: number;
  height: number;
}

export const ASPECT_RATIO_PRESETS: AspectRatioPreset[] = [
  { id: "16:9", name: "16:9 (1080p)", width: 1920, height: 1080 },
  { id: "16:9-720", name: "16:9 (720p)", width: 1280, height: 720 },
  { id: "4:3", name: "4:3 (1024x768)", width: 1024, height: 768 },
  { id: "1:1", name: "1:1 (Square)", width: 1080, height: 1080 },
];

// ===== State Types =====

export interface VideoGenerateState {
  isGenerating: boolean;
  error: string | null;
  result: VideoGenerationResult | null;
  progress: number; // 0-100
  currentSlide: number;
  totalSlides: number;
  slides: SlideContent[];
  config: VideoConfig;
  ttsConfig: Partial<VideoTTSConfig>;
  availability: VideoAvailability | null;
}

File: src-tauri/src/commands/video_generate.rs
//! Video generation Tauri commands — narrated presentations from study content.

use serde::{Deserialize, Serialize};
use std::fs;
use std::sync::mpsc;
use tauri::{AppHandle, Emitter, State};
use uuid::Uuid;

use crate::python_bridge::{SlideContent, VideoGenerationResult};
use crate::AppState;

use super::notebook::CommandError;

/// Progress event payload for video generation
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoProgressPayload {
    pub current_slide: i32,
    pub total_slides: i32,
    pub status: String,
    pub notebook_id: String,
}

/// TTS configuration for video narration
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoTTSConfig {
    pub provider: String,
    pub voice: String,
    pub api_key: Option<String>,
    pub base_url: Option<String>,
    pub model: Option<String>,
    pub speed: Option<f64>,
}

/// Video configuration from the frontend
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoConfig {
    #[serde(default = "default_video_width")]
    pub width: i32,
    #[serde(default = "default_video_height")]
    pub height: i32,
    #[serde(default = "default_theme")]
    pub theme: String, // light, dark
    #[serde(default = "default_transition")]
    pub transition: String, // cut, fade
    pub title: Option<String>,
    pub accent_color: Option<String>, // Custom accent color hex
}

fn default_video_width() -> i32 {
    1920
}

fn default_video_height() -> i32 {
    1080
}

fn default_theme() -> String {
    "light".to_string()
}

fn default_transition() -> String {
    "cut".to_string()
}

/// Generate a narrated video from study content slides
#[tauri::command]
pub async fn generate_study_video(
    app_handle: AppHandle,
    state: State<'_, AppState>,
    notebook_id: String,
    slides: Vec<SlideContent>,
    tts_config: VideoTTSConfig,
    video_config: Option<VideoConfig>,
) -> Result<VideoGenerationResult, CommandError> {
    let nb_id = Uuid::parse_str(&notebook_id).map_err(|e| CommandError {
        message: format!("Invalid notebook ID: {}", e),
    })?;

    if slides.is_empty() {
        return Err(CommandError {
            message: "No slides provided for video generation".to_string(),
        });
    }

    // Get assets directory from storage
    let output_dir = {
        let storage = state.storage.lock().map_err(|e| CommandError {
            message: format!("Failed to acquire storage lock: {}", e),
        })?;

        // Output directory: {notebook_assets}/videos/
        let assets_dir = storage.notebook_assets_dir(nb_id);
        let videos_dir = assets_dir.join("videos");
        fs::create_dir_all(&videos_dir).map_err(|e| CommandError {
            message: format!("Failed to create videos directory: {}", e),
        })?;

        videos_dir
            .to_str()
            .map(|s| s.to_string())
            .ok_or_else(|| CommandError {
                message: "Invalid path encoding".to_string(),
            })?
    };

    // Clone python_ai Arc for spawn_blocking
    let python_ai = state.python_ai.clone();
    let config = video_config.unwrap_or(VideoConfig {
        width: default_video_width(),
        height: default_video_height(),
        theme: default_theme(),
        transition: default_transition(),
        title: None,
    });

    // Create a channel for progress updates
    let (progress_tx, progress_rx) = mpsc::channel::<(i32, i32, String)>();
    let notebook_id_clone = notebook_id.clone();

    // Spawn a task to emit progress events
    let app_handle_clone = app_handle.clone();
    std::thread::spawn(move || {
        while let Ok((current, total, status)) = progress_rx.recv() {
            let payload = VideoProgressPayload {
                current_slide: current,
                total_slides: total,
                status,
                notebook_id: notebook_id_clone.clone(),
            };
            let _ = app_handle_clone.emit("video-generation-progress", payload);
        }
    });

    tauri::async_runtime::spawn_blocking(move || {
        let python_ai = python_ai.lock().map_err(|e| CommandError {
            message: format!("Failed to acquire Python AI lock: {}", e),
        })?;

        python_ai
            .generate_video_with_progress(
                slides,
                &output_dir,
                &tts_config.provider,
                &tts_config.voice,
                tts_config.api_key.as_deref(),
                tts_config.base_url.as_deref(),
                tts_config.model.as_deref(),
                tts_config.speed,
                config.width,
                config.height,
                &config.theme,
                &config.transition,
                config.title.as_deref(),
                progress_tx,
            )
            .map_err(|e| CommandError {
                message: format!("Video generation error: {}", e),
            })
    })
    .await
    .map_err(|e| CommandError {
        message: format!("Task join error: {}", e),
    })?
}

/// Check availability of video generation features
#[tauri::command]
pub fn check_video_generation_availability(
    state: State<AppState>,
) -> Result<serde_json::Value, CommandError> {
    let python_ai = state.python_ai.lock().map_err(|e| CommandError {
        message: format!("Failed to acquire Python AI lock: {}", e),
    })?;

    python_ai
        .check_video_generation_availability()
        .map_err(|e| CommandError {
            message: format!("Failed to check video generation availability: {}", e),
        })
}
